src/main/scala/io/shiftleft/fuzzyc2cpg/FuzzyC2Cpg.scala

```diff
@@ -113,6 +127,52 @@ class FuzzyC2Cpg(outputModuleFactory: CpgOutputModuleFactory) {
 
 }
 
+object FuzzyC2CpgCache {
+  private val functionDeclarations = new mutable.HashMap[String, mutable.ListBuffer[(String, CpgStruct.Builder)]]()
```

Comment:
Is there any reason we're not able to use `ConcurrentHashMap` here?

Reply:
I was considering that, but if I understand correctly, ConcurrentHashMap only ensures that map operations can be executed concurrently without breaking the map, but it does not ensure that multiple operations on a map will be executed without interruption by another thread that wishes to access the map. We need to ensure that both the lookup and corresponding update are atomic, and therefore, I think a lock is required in any case. If we're going to lock operations anyway, then we don't need ConcurrentHashMap anymore - that was my thinking. Let me know if that makes sense.

Reply:
I _think_ you can use the `compute` method on `ConcurrentHashMap` to achieve this, but this is fine :smile_cat: